presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java

```diff
@@ -332,8 +335,44 @@ else if (javaType == Slice.class) {
         @Override
         protected Object visitDereferenceExpression(DereferenceExpression node, Object context)
         {
-            // Dereference is never a Symbol
-            return node;
+            Type type = expressionTypes.get(node.getBase());
+            // if there is no type for the base of Dereference, it must be QualifiedName
+            if (type == null) {
+                return node;
+            }
+
+            Object base = process(node.getBase(), context);
+            if (hasUnresolvedValue(base)) {
+                return new DereferenceExpression(toExpression(base, type), node.getFieldName());
+            }
+
+            RowType rowType = checkType(type, RowType.class, "type");
+            Block row = (Block) base;
+            Type returnType = expressionTypes.get(node);
+            List<RowType.RowField> fields = rowType.getFields();
+            int index = -1;
+            for (int i = 0; i < fields.size(); i++) {
+                RowType.RowField field = fields.get(i);
+                if (field.getName().isPresent() && field.getName().get().equalsIgnoreCase(node.getFieldName())) {
+                    checkArgument(index < 0, "Ambiguous field %s in type %s", field, rowType.getDisplayName());
+                    index = i;
+                }
+            }
+            checkState(index >= 0, "could not find field name: %s", node.getFieldName());
+            if (row.isNull(index)) {
+                return null;
+            }
+            switch (returnType.getJavaType().getSimpleName().toLowerCase()) {
+                case "long":
+                    return returnType.getLong(row, index);
+                case "double":
+                    return returnType.getDouble(row, index);
+                case "boolean":
+                    return returnType.getBoolean(row, index);
+                case "slice":
+                    return returnType.getSlice(row, index);
+            }
+            return returnType.getObject(row, index);
```

Comment:
This is going to break if we add `int` as a native container type, so I would put this in an `else if` and only do it when the native container type is a reference type (other than `Slice`)